numpy/lib/recfunctions.py

```diff
@@ -874,16 +874,33 @@ def _get_fields_and_offsets(dt, offset=0):
     scalar fields in the dtype "dt", including nested fields, in left
     to right order.
     """
+
+    # counts up elements in subarrays, including nested subarrays, and returns
+    # base dtype and count
+    def count_elem(dt):
+        count = 1
+        while dt.shape != ():
+            for size in dt.shape:
+                count *= size
+            dt = dt.base
+        return dt, count
+
     fields = []
     for name in dt.names:
         field = dt.fields[name]
-        if field[0].names is None:
-            count = 1
-            for size in field[0].shape:
-                count *= size
-            fields.append((field[0], count, field[1] + offset))
-        else:
-            fields.extend(_get_fields_and_offsets(field[0], field[1] + offset))
+        f_dt, f_offset = field[0], field[1]
+        f_dt, n = count_elem(f_dt)
+
+        if f_dt.names is None:
+            fields.append((np.dtype((f_dt, (n,))), n, f_offset + offset))
+        else:  # structured dtype
+            subfields = _get_fields_and_offsets(f_dt, f_offset + offset)
+            fields.extend(subfields)
```

Comment:
This line looks wrong to me - doesn't the loop below cover it?

Reply:
Perhaps a pre-optimization, but this way the n=0 case doesn't create a whole new list in the loop below. The idea is the two lines above are the "fast, normal" case, while the loop below account for the "special" case (since I expect subarrays of structures are rare).

Reply:
Ah, I didn't spot the range below starting at 1. I assume by "n=0 case" you meant "n=1 case"?

I think this is a pre-optimization - in particular, it looks to me like you got the handling of `n==0` wrong here as a result of separating the `n` case from the `1` case.

Reply:
A more obviously correct optimization might be:
```
for i in range(n):
    if i == 0:
        # optimization: avoid the list comprehension if it results in no change
        fields.extend(subfields)
    else:
        fields.extend([(d, c, o + i*size) for d, c, o in subfields])
```